src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -888,6 +905,79 @@ internal void ExecuteSubmission(BuildSubmission submission, bool allowMainThread
             }
         }
 
+        /// <summary>
+        /// This method adds the graph build request in the specified submission to the set of requests being handled by the scheduler.
+        /// </summary>
+        internal void ExecuteGraphSubmission(BuildSubmission submission)
+        {
+            lock (_syncLock)
+            {
+                VerifyStateInternal(BuildManagerState.Building);
+
+                try
+                {
+                    // Create the build request
+                    submission.BuildRequest = new BuildRequest(
+                        submission.SubmissionId,
+                        BackEnd.BuildRequest.InvalidNodeRequestId,
+                        GetNewConfigurationId(), // TODO: Should we create a real BuildRequestConfiguration for each graph request?
+                        submission.BuildRequestData.TargetNames,
+                        submission.BuildRequestData.HostServices,
+                        BuildEventContext.Invalid,
+                        null,
+                        submission.BuildRequestData.Flags,
+                        submission.BuildRequestData.RequestedProjectState);
+
+                    if (_shuttingDown)
+                    {
+                        // We were already canceled!
+                        BuildResult result = new BuildResult(submission.BuildRequest, new BuildAbortedException());
+                        submission.CompleteResults(result);
+                        submission.CompleteLogging(true);
+                        CheckSubmissionCompletenessAndRemove(submission);
+                        return;
+                    }
+
+                    if (_graphSchedulingCancellationSource == null)
+                    {
+                        _graphSchedulingCancellationSource = new CancellationTokenSource();
+                    }
+
+                    // Do the scheduling in a separate thread to unblock the calling thread
+                    Task.Factory.StartNew(
+                        () =>
+                        {
+                            try
+                            {
+                                ExecuteGraphBuildScheduler(submission);
+                            }
+                            catch (Exception ex)
```

Comment:
```suggestion
                            catch (Exception ex) when (!ExceptionHandling.IsCriticalException(ex))
```

(and remove the `if` below)